lib/src/stream_transformers/buffer.dart

```diff
@@ -0,0 +1,133 @@
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:async';
+
+/// Creates a [StreamTransformer] which collects values and emits when it sees a
+/// value on [trigger].
+///
+/// If there are no pending values when [trigger] emits, the next value on the
+/// source Stream will immediately flow through. Otherwise, the pending values
+/// are released when [trigger] emits.
+///
+/// Errors from the source stream or the trigger are immediately forwarded to
+/// the output.
+StreamTransformer<T, List<T>> buffer<T>(Stream trigger) =>
+    new _Buffer(trigger, _collectToList);
+
+List<T> _collectToList<T>(T element, List<T> soFar) {
+  soFar ??= <T>[];
+  soFar.add(element);
+  return soFar;
+}
+
+/// A strategy for aggregating values.
+typedef R _Collect<T, R>(T element, R soFar);
+
+/// A StreamTransformer which aggregates values and emits when it sees a value
+/// on [_trigger].
+///
+/// If there are no pending values when [_trigger] emits the first value on the
+/// source Stream will immediately flow through. Otherwise, the pending values
+/// and released when [_trigger] emits.
+///
+/// Errors from the source stream or the trigger are immediately forwarded to
+/// the output.
+class _Buffer<T, R> implements StreamTransformer<T, R> {
+  final Stream _trigger;
+  final _Collect _collect;
+
+  _Buffer(this._trigger, this._collect);
+
+  @override
+  Stream<R> bind(Stream<T> values) {
+    StreamController<R> controller;
+    if (values.isBroadcast) {
+      controller = new StreamController<R>.broadcast();
+    } else {
+      controller = new StreamController<R>();
+    }
+
+    R currentResults;
+    bool waitingForValue = false;
+    StreamSubscription valuesSub;
+    StreamSubscription triggerSub;
+
+    cancelValues() {
+      var sub = valuesSub;
+      valuesSub = null;
+      return sub?.cancel() ?? new Future.value();
+    }
+
+    cancelTrigger() {
+      var sub = triggerSub;
+      triggerSub = null;
+      return sub?.cancel() ?? new Future.value();
+    }
+
+    closeController() {
+      var ctl = controller;
+      controller = null;
+      return ctl?.close() ?? new Future.value();
+    }
+
+    emit() {
+      controller.add(currentResults);
+      currentResults = null;
+    }
+
+    onValue(T value) {
+      currentResults = _collect(value, currentResults);
+      if (waitingForValue) {
+        emit();
+        waitingForValue = false;
+      }
+    }
+
+    valuesDone() {
+      valuesSub = null;
+      if (currentResults == null) {
+        closeController();
+        cancelTrigger();
+      }
+    }
+
+    onTrigger(_) {
+      if (currentResults == null) {
+        waitingForValue = true;
+        return;
+      }
+      emit();
+      if (valuesSub == null) {
+        closeController();
+        cancelTrigger();
+      }
+    }
+
+    triggerDone() {
+      cancelValues();
+      closeController();
+    }
+
+    controller.onListen = () {
+      valuesSub = values.listen(onValue,
```

Comment:
I don't think its super important to support, users can always call `asBroadcastStream` on the resulting stream right?